tensorboard/plugins/graph/README.md

```diff
@@ -121,9 +121,384 @@ describing the motivations, research, and process behind it.
+
+    -   Add edges to SlimGraph
+
+        -   Detect reference edges using a hardcoded list (e.g. Assign,
+            AssignAdd).
+        -   Move the inputs of the (Const) in-embedding node into incoming edges
+            of the main node.
+        -   Move the inputs of the (*Summary) out-embedding node into inputs of
+            the main node where the out-embedding points to.
+
+    -   Returns
+
+    ```
+    {
+      nodes: {
+        "adder/Add": OpNode,
+        ...
+      };
+      edges: BaseEdge[],
+      ...
+    }
+    ```
+
+### Building the Hierarchy
+
+View components create the RenderGraph mainly from the Hierarchy. As we will
+see, the key difference between a Hierarchy and a SlimGraph is that a Hierarchy
+contains more nodes beyond the OpNodes: MetaNodes are created to group other
+nodes, and MetaEdges are a layer on top of the SlimGraph's BaseEdges.
+
+Furthermore, the Hierarchy building phase will compute expensive information
+about the graph and store it on the data structure, such as templates. Template
+ids are stored on MetaNodes, and are used by the view components to determine
+node coloring. When 'Color by structure' is enabled, MetaNodes with similar
+structure are assumed to have the same templateId, and thus assigned the same
+color.
+
+The BridgeGraph is not computed at this phase, but instead on-demand by the view
+components.
+
+Note: building a Hierarchy modifies the SlimGraph's OpNodes in place!
+
+**tf_graph_common/hierarchy.ts: build(SlimGraph): Hierarchy**
+
+-   Create an artificial MetaNode for the root.
+-   Create an index of nodes as `Record<nodeId: string,
+    OpNode|SeriesNode|MetaNode>`
+-   'Adding nodes'
+    -   Iterate over OpNodes, keep track of device names, XLAClusters,
+        opHistograms, TPU Compatibility histograms, library function nodes and
+        their associated function name, etc.
+    -   Assume the node's name's '/' occurences indicate its level from the
+        root. If node is named 'a/b/c', create metanodes 'a' and 'a/b'.
+    -   Update embedding nodes to point to the OpNode they are embedded to:
+        `embedding.parentNode = node;`.
+-   'Detect series'
+    -   Recursively traverse metanodes starting from the top-level. At each
+        metanode,
+        -   Take the direct children nodes and group them into clusters by
+            op-type.
+        -   For each op cluster, detect series by greedily looking for numeric
+            suffix changes. For example, 3 OpNodes with op 'Add': `add_1`,
+            `add_2`, `add_3` will be grouped into a SeriesNode 'add_[1-3]'.
+        -   Series groups require at least 2 nodes (later we filter out groups
+            with less than 5 nodes) where the numeric suffix is in increasing
+            order.
+        -   Children within a SeriesNode have a reference: `child.owningSeries =
+            seriesName`.
+    -   If series has 5+ members,
```

Comment:
I asked myself this question reading section above. Perhaps call this out before L329?

Reply:
Updated the above line to say 5, which is a configurable parameter.